server/src/main/java/io/deephaven/server/plugin/type/ObjectTypes.java

```diff
@@ -0,0 +1,85 @@
+package io.deephaven.server.plugin.type;
+
+import io.deephaven.plugin.type.ObjectType;
+import io.deephaven.plugin.type.ObjectTypeRegistration;
+import io.deephaven.plugin.type.ObjectTypeClassBase;
+import io.deephaven.plugin.type.ObjectTypeLookup;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import javax.lang.model.SourceVersion;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * Provides synchronized object type {@link ObjectTypeRegistration registration} and {@link ObjectTypeLookup lookup}.
+ *
+ * <p>
+ * Object type registration that is an instances of {@link ObjectTypeClassBase} receives special consideration, and
+ * these objects have more efficient lookups.
+ */
+@Singleton
+public final class ObjectTypes implements ObjectTypeLookup, ObjectTypeRegistration {
+
+    private static final Set<String> RESERVED_TYPE_NAMES_LOWERCASE = Set.of("table", "tablemap", "treetable", "");
```

Comment:
should we reserve `figure` in some way? not sure of the mechanics of how this works, if it makes any more sense to reserve tablemap than figure...

Reply:
I'm more inclined to want to namespace Figure as something like `io.deephaven.Figure` (doesn't have to be the full Figure class, but something that could be stable and survive implementation refactorings). As it is now, if somebody else tries to install a "Figure" object type, we'll blow up at startup.

Reply:
Should verify that "tablemap" is going to be a thing (with that word) - but while i agree that our "table" should win over other tables, i'm not sure why our tablemap or treetable would be any more essential than our figures when it comes to serialization to a client.

Reply:
Assuming we can revisit this later.